tsl/src/nodes/decompress_chunk/batch_queue_heap.c

```diff
@@ -175,65 +181,125 @@ batch_queue_heap_push_batch(DecompressChunkState *chunk_state, TupleTableSlot *c
 		return;
 	}
 
-	chunk_state->merge_heap =
-		binaryheap_add_unordered_autoresize(chunk_state->merge_heap, new_batch_index);
+	bqh->merge_heap = binaryheap_add_unordered_autoresize(bqh->merge_heap, new_batch_index);
 }
 
-TupleTableSlot *
-batch_queue_heap_top_tuple(DecompressChunkState *chunk_state)
+static TupleTableSlot *
+batch_queue_heap_top_tuple(BatchQueue *bq)
 {
-	BatchArray *batch_array = &chunk_state->decompress_context.batch_array;
+	BatchQueueHeap *bqh = (BatchQueueHeap *) bq;
+	BatchArray *batch_array = &bq->batch_array;
 
-	if (binaryheap_empty(chunk_state->merge_heap))
+	if (binaryheap_empty(bqh->merge_heap))
 	{
 		return NULL;
 	}
 
-	const int top_batch_index = DatumGetInt32(binaryheap_first(chunk_state->merge_heap));
+	const int top_batch_index = DatumGetInt32(binaryheap_first(bqh->merge_heap));
 	DecompressBatchState *top_batch = batch_array_get_at(batch_array, top_batch_index);
 	Assert(!TupIsNull(top_batch->decompressed_scan_slot));
 	return top_batch->decompressed_scan_slot;
 }
 
-void
-batch_queue_heap_create(DecompressChunkState *chunk_state)
+static void
+batch_queue_heap_reset(BatchQueue *bq)
 {
-	DecompressContext *dcontext = &chunk_state->decompress_context;
-
-	batch_array_init(&dcontext->batch_array,
-					 INITIAL_BATCH_CAPACITY,
-					 chunk_state->num_compressed_columns,
-					 dcontext->batch_memory_context_bytes);
+	BatchQueueHeap *bqh = (BatchQueueHeap *) bq;
+	binaryheap_reset(bqh->merge_heap);
+}
 
-	chunk_state->merge_heap = binaryheap_allocate(INITIAL_BATCH_CAPACITY,
-												  decompress_binaryheap_compare_heap_pos,
-												  chunk_state);
+/*
+ * Free the binary heap.
+ */
+static void
+batch_queue_heap_free(BatchQueue *bq)
+{
+	BatchQueueHeap *bqh = (BatchQueueHeap *) bq;
+	BatchArray *batch_array = &bq->batch_array;
 
-	Assert(chunk_state->last_batch_first_tuple == NULL);
-	chunk_state->last_batch_first_tuple =
-		MakeSingleTupleTableSlot(chunk_state->csstate.ss.ss_ScanTupleSlot->tts_tupleDescriptor,
-								 &TTSOpsVirtual);
+	elog(DEBUG3, "Heap has capacity of %d", bqh->merge_heap->bh_space);
+	elog(DEBUG3, "Created batch states %d", batch_array->n_batch_states);
+	batch_array_clear_all(&bq->batch_array);
+	binaryheap_free(bqh->merge_heap);
+	bqh->merge_heap = NULL;
+	ExecDropSingleTupleTableSlot(bqh->last_batch_first_tuple);
+	batch_array_destroy(batch_array);
 }
 
-void
-batch_queue_heap_reset(DecompressChunkState *chunk_state)
+const struct BatchQueueFunctions BatchQueueFunctionsHeap = {
+	.free = batch_queue_heap_free,
+	.needs_next_batch = batch_queue_heap_needs_next_batch,
+	.pop = batch_queue_heap_pop,
+	.push_batch = batch_queue_heap_push_batch,
+	.reset = batch_queue_heap_reset,
+	.top_tuple = batch_queue_heap_top_tuple,
+};
+
+static SortSupport
+build_batch_sorted_merge_info(const List *sortinfo, int *nkeys)
 {
-	binaryheap_reset(chunk_state->merge_heap);
+	if (sortinfo == NIL)
+	{
+		*nkeys = 0;
+		return NULL;
```

Comment:
Probably can  be an assertion.

Reply:
I just moved this code and didn't change it so I don't fully know the reasoning behind it. But I will change to an assertion.

Reply:
If we do not need to support `nkeys == 0` then we can raise an error here. I would actually suggest using `Ensure` for the odd case that `nkeys` happen to be instead of silently ignoring it. Although I think the code currently support `nkeys == 0` I think it might break in the future (in release builds) if our testing contains the assertion.